### ISSUE TITLE
API for getting caret selection

### ISSUE THREAD

**Issue by josdejong** [josdejong/jsoneditor#470]
@meirotstein has implemented a status bar where you can see the caret location. A discussion popped up there (https://github.com/josdejong/jsoneditor/issues/443#issuecomment-331626325) to create an API to get the current selection.

I think this API will contain a method `getSelection` and a callback `onSelectionChange`, and it probably needs a different implementation for the modes `tree`/`form`/`view` vs `text`/`code`.

**Comment by meirotstein** [comment]
IMHO - I think that for `text` and `code` the callback should return the selected text and cursor positions (row/col for start and end)

As for `tree`, `form` and `view` - here the selected node should be returned, so the callback 
signature is different.

Perhaps it should be separated into different functions to avoid confusions? also I'm not sure that a `get` function is the most recommended here as it usually means a single value and not a registration of a listener, what about the following suggestions:

```
//text and code
onTextSelectionChanged(function(text, startPos, endPos){
...
})

//tree, form and view
onNodeSelectionChanged(function(node){
...
})
```

**Comment by josdejong** [comment]
> Perhaps it should be separated into different functions to avoid confusions?

Yes I think so.

I like your proposal for `onTextSelectionChanged` and `onNodeSelectionChanged`. Instead of passing `node` to `onNodeSelectionChanged` though I think we can pass the *path* of the current node. It is also possible to have a multi-selection, so maybe the syntax should have a start and an end, something like:
```js
//tree, form and view
onNodeSelectionChanged(function(startPath: string[], endPath: string[]) {
...
})
```
Maybe there is also need to know exactly where the cursor is in the node: `field` or `value`, and maybe even the cursor position in the editable div, then we need to pass more information, something like:
```js
//tree, form and view
onNodeSelectionChanged(function(startNode: Object, endNode: Object) {
...
})
// where startNode and endNode have a structure:
// {
//   path: string[],
//   selection: {
//     input: 'field' | 'value'
//     start: number,  // start character index 
//     end: number     // end character index
//   }
// } 
```
Ideally the API should be a bit similar to for example `onEditable` which also passes a "node" object.

**Comment by YuriyChaban** [comment]
josdejong, did you implement functionality for getting caret selection in API?
I need help with getting caret position and when click on some button in place where caret located, some json adding in that location in CODE editor

LIke this example:

http://i.prntscr.com/mKblu6WESCSUh1Vu4iCcng.png

if you have any idea how can i do this, please write me. Thank you.

**Comment by josdejong** [comment]
Getting caret position isn't yet implemented. In case of the `code` mode, you could use something like `codeEditor.aceEditor.getCursorPosition()`.

**Comment by meirotstein** [comment]
Now that the caret position is available for both text and code I can add the API quite easily - hope to do that in the in the following days.
As for `onNodeSelectionChanged` - @josdejong , you suggested:

> onNodeSelectionChanged(function(startNode: Object, endNode: Object)

Perhaps it will be more clear to send a list of the selected node instead of start and end? also, what about inner node - should they also be considered as selected in that case or only to send the higher parent that is selected - IMHO this is what we should do, but maybe I'm missing something here.

**Comment by josdejong** [comment]
Thanks Meir! Yeah I think for text and code mode it's a piece of cake now. Good to think about the implementation of `onNodeSelectionChanged` a bit deeper. I think these functions `onNodeSelectionChanged` and `onTextSelectionChanged` could/should become complementary to two functions `onNodeSelection` and `setTextSelection`. Since a selection is a range from a start to an end, that's what the selection objects should hold. 

In practice, you may want to do "something" with all selected nodes, so you want an easy way to list or iterate over all nodes in the selection. We could provide a utility function for that separately, something like `getSelectedPaths(selection)` or `getSelectedNodes(selection)`, or simply `getSelectedPaths()`, or a static util function `getPaths(json, selection)`, ...

What do you think?

**Comment by YuriyChaban** [comment]
Thanks for explenation.

**Comment by YuriyChaban** [comment]
Hello jodejong, can you help me with understanding in editor, how can i update all editor when i paste a json with a button? I mast update `frame `or smth else, becouse when i click button, json pasted but not showing and when i keydown some button from keyboard, pasted json is showed becouse editor updated after keydown event. But when i try to understand for what element i must update view i can't do this, because i cant't find element what i need for updating editor.

**Comment by josdejong** [comment]
@YuriyChaban can you please ask your question in a new topic, so we keep this one focused on discussing an API for getting a selection?

**Comment by YuriyChaban** [comment]
yes, of course

**Comment by iRuisu** [comment]
Apologies for how hacky this is, but if you're looking for a quick solution until this feature has officially been implemented you can do the following edits to your local copy of JSON Editor to get the JSON path from the Tree/Form modes.

Within your local copy of **jsoneditor.js** add these two lines of code below **this.treePath.setPath(pathObjs)**:

```javascript
this.treePath.pathObjs = pathObjs;
this.treePath.pathNodes = pathNodes;
```

You can then use this function to get the JSON path of the selected attribute:

```javascript
function getJSONPath( editor ) {
	var 	pathObjs = editor.treePath.pathObjs,
		pathNodes = editor.treePath.pathNodes,
		jsonPath = []
	
	for( var i = 0; i < pathObjs.length; i++ ) {
		
		if( pathObjs[i].node.field != undefined ) {
		
			jsonPath.push( pathObjs[i].node.field )
			
		} else if( pathNodes[i].index ) {
		
			jsonPath.push( "[" + pathNodes[i].index + "]" )
			
		}
		
	}
	
	return jsonPath.join( "." ).replace( ".[" , "[" )
}
```

The **editor** parameter is the instance of your editor.

Can't wait to see this officially implemented as this is an awesome tool!

**Comment by meirotstein** [comment]
@josdejong 

>  I think these functions onNodeSelectionChanged and onTextSelectionChanged could/should become complementary to two functions onNodeSelection and setTextSelection.

If I understand you correctly you suggest to provide a setter for selections in addition to an async getter - this makes sense.

However, in regard to `setNodeSelection` - selections of nodes from different levels can be tricky, as the editor currently limits to selection to nodes from the same level as far as I managed to see.. anyhow I'm not sure how it will behave with various of start/end selection configurations.. my suggestion is to limit this api for a single selection - at least in the scope of this feature.

As for the utility function - I'm not sure this would be necessary as once you have the selected node in your hand you can go over it's children quite easily - it is eventually a simple BFS/DFS implementation.

So to summarize, this addition will includes 4 new api methods:

- `onNodeSelectionChanged(startNode, endNode))`
- `onTextSelectionChanged(function(text, startPos, endPos))`
- `setNodeSelection(node)`
- `setTextSelection(startPos, endPos)`

@iRuisu - this is a nice workaround however this is not exactly the outcome of this feature - your code actually provides the path from parent to the selection (which is what the treePath projects), while in this feature we aim to provide the selections themselves.
Of course - once this additions will be available it will be very handy for your needs as you'll be able to get the selected node and call for 'getNodePath()'

**Comment by josdejong** [comment]
@meirotstein when the selected start and end node are not on the same level, what JSONEditor does right now is find the common parent of the two, and select the childs in the range of the start/end of this parent, see [`treemode._findTopLevelNodes `](https://github.com/josdejong/jsoneditor/blob/master/src/js/treemode.js#L1114).

You may be right, maybe we do not need a util function at all.

Your summary of the 4 new API methods is correct, though I think we should add two more: `getNodeSelection` and `getTextSelection`, so you can get the current state at any arbitrary moment (like when a user presses a Save button) without having to listen for changes.

**Comment by meirotstein** [comment]
@josdejong - got it. hopefully I will deliver it soon.

**Comment by josdejong** [comment]
Looking forward to it :)

**Comment by meirotstein** [comment]
@josdejong I think this issue is considered as closed :)

**Comment by josdejong** [comment]
LOL, you're right
